### PR TITLE
Implement `SMO` instruction

### PR DESCRIPTION
**Pull request by vlopes11**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -12,11 +12,13 @@ description = "FuelVM interpreter."
 
 [dependencies]
 dyn-clone = { version = "1.0", optional = true }
-fuel-asm = "0.5"
+#fuel-asm = "0.6"
+fuel-asm = { git = "https://github.com/FuelLabs/fuel-asm", branch = "vlopes11/gtf" }
```
Comment (adlerjohn) [inline Cargo.toml:16]: Unblocked?

File: src/interpreter/executors/instruction.rs [hunk 2]
```diff
@@ -240,16 +242,6 @@ where
                 self.alu_set(ra, b ^ imm)?;
             }
 
-            OpcodeRepr::CIMV => {
```
Comment (Voxelot) [inline src/interpreter/executors/instruction.rs:243]: ideally we remove these in a separate PR to make it easier to track what changes were made in each release.

Comment (vlopes11) [reply src/interpreter/executors/instruction.rs:243]: We have this pr: https://github.com/FuelLabs/fuel-vm/pull/158

But its blocked by the fuel-tx one: https://github.com/FuelLabs/fuel-tx/pull/149

File: src/interpreter/blockchain.rs [hunk 3]
```diff
@@ -308,4 +309,64 @@ where
...
+            if self.is_external_context() {
+                self.base_asset_balance_sub(amount)?;
+            } else {
+                self.external_asset_balance_sub(amount)?;
+            }
+        }
+
+        let fp = self.registers[REG_FP] as usize;
+        let txid = self.tx_id();
+        let data = a as usize + Address::LEN;
```
Comment (pixelcircuits) [inline src/interpreter/blockchain.rs:357]: Maybe this should use `Bytes32::LEN` like on line 316 for consistency.

File: src/interpreter/executors/instruction.rs [hunk 4]
```diff
@@ -36,11 +36,13 @@ where
         // Should execute `lo` only if there is no rupture in the flow - that means
         // either a breakpoint or some instruction that would skip `lo` such as
         // `RET`, `JI` or `CALL`
-        if self.registers[REG_PC] == pc && state.should_continue() {
+        let result = if self.registers[REG_PC] == pc && state.should_continue() {
```
Comment (pixelcircuits) [inline src/interpreter/executors/instruction.rs:39]: Why was `result` added?

File: src/interpreter/blockchain.rs [hunk 5]
```diff
@@ -308,4 +309,64 @@ where
...
+        }
+
+        // validations passed, perform the mutations
+
+        // subtract the balance only if there is an amount set
+        if amount > 0 {
+            if self.is_external_context() {
+                self.base_asset_balance_sub(amount)?;
+            } else {
+                self.external_asset_balance_sub(amount)?;
```
Comment (pixelcircuits) [inline src/interpreter/blockchain.rs:351]: The SMO amount only ever refers to the base asset. In ether context, `self.base_asset_balance_sub` should be used

File: src/interpreter/internal.rs [hunk 6]
```diff
@@ -155,6 +155,20 @@ impl<S> Interpreter<S> {
         Ok(())
     }
 
+    /// Reduces the unspent balance of the asset defined in $fp
+    pub(crate) fn external_asset_balance_sub(&mut self, value: Word) -> Result<(), RuntimeError> {
+        let fp = self.registers[REG_FP] as usize;
+        let asset_id = &self.memory[fp..fp + AssetId::LEN];
```
Comment (pixelcircuits) [inline src/interpreter/internal.rs:161]: Isn't the `asset_id` the second set of 32 bytes on the [call frame](https://github.com/FuelLabs/fuel-specs/blob/master/specs/vm/main.md#call-frames)?

Comment (adlerjohn) [reply src/interpreter/internal.rs:161]: There should be a test for this is what this means!

File: src/interpreter/blockchain.rs (hunk 3 above)
Comment (vlopes11) [reply src/interpreter/blockchain.rs:357]: True, they should be the same. The concrete type is in fact `Address`, instead of raw bytes. Better to use that so its clear what is expected of this register:

https://github.com/FuelLabs/fuel-vm/pull/159/commits/7bd578949cb2d81e1bd84388532d33e523473636

File: src/interpreter/executors/instruction.rs (hunk 4 above)
Comment (vlopes11) [reply src/interpreter/executors/instruction.rs:39]: This is to facilitate debug and readability. If we just return complex expressions without assigning to a variable, then we need to inspect the frame pointers in gdb - having an assigned variable is zero runtime overhead and will give us that readability for free

File: src/interpreter/blockchain.rs (hunk 5 above)
Comment (vlopes11) [reply src/interpreter/blockchain.rs]: https://github.com/FuelLabs/fuel-vm/pull/159/commits/4398216d539b52c63b2098ea4ddf38245b480924

Do you think we should rephrase here to simplify the description since we only update the base asset?

https://github.com/FuelLabs/fuel-specs/blob/bd6ec935e3d1797a192f731dadced3f121744d54/specs/vm/instruction_set.md?plain=1#L1417

File: src/interpreter/internal.rs (hunk 6 above)
Comment (vlopes11) [reply src/interpreter/internal.rs]: This function was added as a specific use-case for SMO and is no longer present

File: src/interpreter/blockchain.rs (hunk 5 above)
Comment (adlerjohn) [reply src/interpreter/blockchain.rs]: That line does say `balanceOfStart(0)` and `asset ID 0`, so it's explicit and unambiguous if you know what you're looking for. More generally, the specs aren't approachable _unless_ you've read them almost entirely, and that probably should be improved.

File: src/interpreter/initialization.rs [hunk 7]
```diff
@@ -88,20 +88,25 @@ impl<S> Interpreter<S> {
 
     // compute the initial free balances for each asset type
     pub(crate) fn initial_free_balances(&self) -> Result<HashMap<AssetId, Word>, InterpreterError> {
+        let base_asset = AssetId::default();
```
Comment (adlerjohn) [inline src/interpreter/initialization.rs:91]: Orthogonal to this PR, but I'm thinking the base asset's ID should be a parameter in the VM. It's not necessarily guaranteed to be `AssetId::default()` always. Both the specs and the implementation should be modular in this regard. Can you file an issue to track changing?

File: src/interpreter/initialization.rs [hunk 8]
```diff
@@ -124,19 +129,20 @@ impl<S> Interpreter<S> {
...
                     })?;
+
+            *base_asset_balance = deducted_balance;
         }
 
-        // reduce free balances by coin and withdrawal outputs
+        // reduce free balances by coin
         for (asset_id, amount) in self.tx.outputs().iter().filter_map(|output| match output {
             Output::Coin { asset_id, amount, .. } => Some((asset_id, amount)),
-            Output::Message { .. } => todo!(),
```
Comment (adlerjohn) [inline src/interpreter/initialization.rs:139]: Wait, message outputs still count as "withdrawals," or rather, they should still reduce the free balance. What is this change doing? Is the output amount accounted for?

File: src/interpreter/transaction.rs [hunk 9]
```diff
@@ -52,7 +52,7 @@ impl<S> Interpreter<S> {
             .tx
             .witnesses()
             .get(b as usize)
-            .ok_or(PanicReason::OutputNotFound)
+            .ok_or(PanicReason::WitnessNotFound)
```
Comment (adlerjohn) [inline src/interpreter/transaction.rs:55]: This looks like an orthogonal bugfix. Ideally make a separate PR for this.

**Review by adlerjohn — COMMENTED** [review]
minor comments

File: src/interpreter/executors/instruction.rs (hunk 4 above)
Comment (pixelcircuits) [reply src/interpreter/executors/instruction.rs:39]: cool!

File: src/interpreter/initialization.rs (hunk 8 above)
Comment (pixelcircuits) [reply src/interpreter/initialization.rs:139]: OutputMessage amount is 0 or the whole transaction is invalid.

Comment (adlerjohn) [reply src/interpreter/initialization.rs:139]: Ah right they're variable.

**Review by adlerjohn — APPROVED** [review]
Looks fine to me, would like @pixelcircuits and @Voxelot to give a review as well.

**Review by Dentosal — APPROVED** [review]

**Review by pixelcircuits — APPROVED** [review]

File: src/interpreter/transaction.rs (hunk 9 above)
Comment (vlopes11) [reply src/interpreter/transaction.rs:55]: Thats true, I just spotted this nit while debugging
